### PR TITLE
Tighten wire-boundary validation rules

### LINKED ISSUE #77: [Tracking] Standardize wire-boundary validation and canonical encoding rules
## Summary

Several small issues point at one recurring elegance problem: different protocol boundaries validate and canonicalize similar wire concepts differently. Some paths enforce single-tag cardinality; others take the first matching tag. Some identifiers are trusted before verification for telemetry. Some internal encodings concatenate fields without a reusable length-delimited tuple encoding.

This issue tracks a shared validation/canonicalization cleanup so every parser and boundary uses the same primitives.

## Child issues / concrete symptoms

- marmot-protocol/mdk#336 — gift-wrap recipient `p` tag uses first-match behavior instead of strict single-tag enforcement, unlike the group `h` tag path.
- marmot-protocol/mdk#350 — welcome rumor `e` / `relays` tags use first-match behavior instead of single-tag enforcement.
- marmot-protocol/mdk#351 — inbound event self-reported `id` is copied into transport `MessageId` / telemetry before id/signature verification.
- marmot-protocol/mdk#349 — `build_key` concatenates label + key + version without length delimiters, same general class as earlier undelimited-concatenation hazards.

## Proposed protocol / design shape

Define reusable boundary rules and helpers:

- **Tag cardinality helpers**: `required_single_tag`, `optional_single_tag`, `repeated_tag_list`, each with explicit duplicate behavior.
- **Strict-by-default parsing** for authenticated routing tags: duplicate singleton tags are rejected, not first-match accepted.
- **Verified identity boundary**: event ids, sender ids, member ids, and routing ids should not be used for durable telemetry/audit/routing decisions until the relevant hash/signature/authentication check has passed, or they must be explicitly marked unverified.
- **Canonical tuple encoding**: any signed bytes, digest inputs, storage keys, or cross-version identifiers should use length-delimited / typed tuple encoding rather than ad hoc concatenation.
- **One parser policy table** in the spec for each Marmot/Nostr event kind and inner app-event kind: required tags, optional tags, repeatable tags, duplicate handling, and canonicalization.

## Acceptance criteria

- The spec documents tag cardinality and duplicate handling for all Marmot-owned event/tag surfaces.
- Gift-wrap `p`, group `h`, welcome `e`, welcome `relays`, and similar singleton tags all use the same helper/policy.
- Unverified event ids are not exposed as trusted `MessageId` values in telemetry/audit; either verify first or label them as untrusted wire ids.
- New signed/digested/stored tuple encodings are length-delimited or otherwise collision-resistant by construction.
- Tests include duplicate singleton tags, repeated list tags, forged event ids, and boundary cases for canonical tuple encodings.

## Why this should be tracked holistically

Each child issue is low-to-medium severity alone, but together they create needless parser diversity. Standardizing the boundary rules removes special cases, makes the spec easier to implement independently, and reduces future review burden.

## Additional child issues (2026-06-30 review pass)

- marmot-protocol/mdk#398 — `group_id_from_hex` skips the 32-byte length check its sibling helpers enforce
- marmot-protocol/mdk#386 — inbound KeyPackages are never validated for lifetime/expiry *(partial / related)*

---
_Generated by [Claude Code](https://claude.ai/code)_

### PR DESCRIPTION
**Pull request by erskingardner**
## Summary

- define unambiguous tuple/preimage boundaries for Marmot-owned signed, hashed, stored, compared, and named bytes
- pin the KeyPackage lifetime acceptance rule to 84 days plus a one-hour skew margin
- add a Nostr tag-cardinality table covering group `h`, gift-wrap `p`, welcome `e`/`relays`, and KeyPackage tags
- state that unverified Nostr event ids and metadata are not trusted routing/replay/telemetry evidence

## Companion implementation

- marmot-protocol/mdk#735 closes the remaining open child issues: mdk#349, mdk#386, and mdk#398.

## Verification

- rg -n "the spec should|spec MUST|Spec-Defined|GroupEvent|\bengine\b|darkmatter" .
- rg -n "PendingStateRef|drain_auto_publish|confirm_published|publish_failed" .
- rg -n "relay hints|internal version|public registry" .

The remaining matches are existing AGENTS/implementation-model guidance and existing push-notification relay-hint text.

Closes #77

---

<a href="https://stagereview.app/marmot-protocol/marmot/pull/236">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

## Summary by CodeRabbit

* **Documentation**
  * Clarified rules for encoding multi-field values so boundaries remain unambiguous and collision-resistant.
  * Tightened KeyPackage validity guidance, including required lifetime checks and rejection conditions.
  * Strengthened Nostr event validation rules, including authenticated metadata, required tag counts, duplicate-tag handling, and pre-processing checks.

### CODE REVIEW

**Comment by ghost** [comment]
Ready to review this PR? Stage has broken it down into 3 individual chapters for you:

| | Title |
|---|-------|
| 1 | [Define unambiguous tuple and preimage boundaries](https://stagereview.app/marmot-protocol/marmot/pull/236/chapters/1) |
| 2 | [Tighten KeyPackage lifetime validation rules](https://stagereview.app/marmot-protocol/marmot/pull/236/chapters/2) |
| 3 | [Enforce Nostr tag cardinality and verification](https://stagereview.app/marmot-protocol/marmot/pull/236/chapters/3) |

<a href="https://stagereview.app/marmot-protocol/marmot/pull/236">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://stagereview.app/assets/gh-open-in-stage-dark.svg">
    Open in Stage
  </picture>
</a>

<sub>Chapters generated by <a href="https://stagereview.app">Stage</a> for commit e5bf36000ca357db52f43fa40c2d42965b662607 on Jul 4, 2026 8:26pm UTC.</sub>

File: transports/nostr.md [hunk 1]
```diff
@@ -54,6 +54,36 @@ Producers SHOULD use `wss`, lowercase DNS hostnames, omit default ports, and avo
...
+| kind `30443` KeyPackage | `mls_ciphersuite` | exactly one id-list tag |
+| kind `30443` KeyPackage | `mls_extensions` | exactly one id-list tag |
+| kind `30443` KeyPackage | `mls_proposals` | exactly one id-list tag |
+| kind `30443` KeyPackage | `app_components` | exactly one id-list tag |
+
+The kind `444` `e` tag is a claim about which KeyPackage event was consumed. It is not proof that the event exists, was
+authored by the invitee account, or carried the decoded KeyPackage. A client that needs those facts MUST fetch and
+verify the referenced kind `30443` event and then verify the decoded KeyPackageRef under
+[../foundation/key-packages.md](../foundation/key-packages.md).
+
```
[the remark this answers is not in the document]
Comment (erskingardner) [reply transports/nostr.md:86]: Addressed the valid part in e5bf360 by marking the event identity/cardinality trust-boundary rules as **CRITICAL** and naming the kind 444 trust boundary explicitly. I did not add a threat_model.md link because this repository does not currently contain a threat_model.md file, and adding a new top-level spec surface just for this follow-up would exceed the scope of this PR.
